p2p/swarm.go

```diff
@@ -877,3 +871,77 @@ func (s *swarm) hasOutgoingPeer(peer p2pcrypto.PublicKey) bool {
 	s.outpeersMutex.RUnlock()
 	return ok
 }
+
+var listeningIp = inet.ParseIP("0.0.0.0")
+
+func (s *swarm) getListeners(
+	getTcpListener func(tcpAddr *inet.TCPAddr) (inet.Listener, error),
+	getUdpListener func(udpAddr *inet.UDPAddr) (net.UDPListener, error),
+	discoverUpnpGateway func() (nat_traversal.UpnpGateway, error),
+) (inet.Listener, net.UDPListener, error) {
+
+	port := s.config.TCPPort
+	randomPort := port == 0
+	var gateway nat_traversal.UpnpGateway
+	if s.config.AcquirePort {
+		var err error
+		gateway, err = discoverUpnpGateway()
+		if err != nil {
+			s.logger.With().Warning("could not discover UPnP gateway", log.Err(err))
+		}
+	}
+
+	for {
+		tcpAddr := &inet.TCPAddr{IP: listeningIp, Port: port}
+		tcpListener, err := getTcpListener(tcpAddr)
+		if err != nil {
+			return nil, nil, fmt.Errorf("failed to acquire requested tcp port: %v", err)
+		}
+
+		addr := tcpListener.Addr()
+		port = addr.(*inet.TCPAddr).Port
+		udpAddr := &inet.UDPAddr{IP: listeningIp, Port: port}
+		udpListener, err := getUdpListener(udpAddr)
+		if err != nil {
+			if err := tcpListener.Close(); err != nil {
+				s.logger.With().Error("error closing tcp listener", log.Err(err))
+			}
+			if randomPort {
+				port = 0
+				continue
+			}
+			return nil, nil, fmt.Errorf("failed to acquire requested udp port: %v", err)
+		}
+
+		if gateway != nil {
```

Comment:
why not move the above part that checks `AquirePort` and tries to find the gateway to here ?

Reply:
Discovering a gateway is slow - it can take up to 30 seconds - I put it outside the loop so it only happens once.